### PR TITLE
Simpler check-typo

### PR DESCRIPTION
**Pull request by gasche**
I wanted this to fix [this particular comment](https://github.com/ocaml/ocaml/pull/1905#discussion_r202782208) of @dra27 in the review of #1905. If the present PR is accepted, I will rebase #1905 on top of it.

### CODE REVIEW

**Comment by gasche** [comment]
@dra27 where is the use of check-typo (and maybe the pre-commit-hook) documented? If you have committed any documentation in the repository (I could only find a mention in passing in CONTRIBUTING.md), I think that it would be nice to have something in HACKING.adoc -- either the toplevel one, or a new tools/HACKING.adoc.

**Comment by dra27** [comment]
@gasche - I'm slightly embarrassed to say that while it's on my TODO list, documentation beyond the mention in CONTRIBUTING is all that I've committed so far.

File: .gitattributes [hunk 1]
```diff
@@ -16,8 +16,8 @@
 # Default behaviour, for if core.autocrlf isn't set
 * text=auto
 
-# Don't believe there's a way to wrap lines in .gitattributes
-.gitattributes ocaml-typo=long-line
+# Don't believe there's a way to wrap lines in .gitattributes files
+.gitattributes typo.long-line=may
```
Comment (dra27) [inline .gitattributes:20]: Nit: if we're renaming `ocaml-typo`, perhaps `lint` is better (and still only 4 chars)?

File: .gitattributes [hunk 2]
```diff
@@ -38,68 +38,81 @@
...
+*.md                     typo.missing-header
+README*                  typo.missing-header
+*.adoc                   typo.missing-header
+*.adoc                   typo.long-line=may
+
+/.mailmap                typo.long-line typo.missing-header typo.non-ascii
+/.merlin                 typo.missing-header
+/Changes                 typo.utf8 typo.missing-header
+/INSTALL                 typo.missing-header
+/LICENSE                 typo.long-line typo.very-long-line typo.missing-header
```
Comment (dra27) [inline .gitattributes]: Surely at this point we could try to subsume long-line and very-long-line into something neater? At the very least, couldn't the Awk script automatically add when `long-line` when `very-long-line` is specified?

**Comment by gasche** [comment]
@dra27: I'm hoping that you will eventually review this PR (and later #1905). There is nothing urgent about it, so feel free to prioritize other things.

**Comment by dra27** [comment]
I had started to review it! I think the changes are great, just that going through it was taking time - I'll definitely come to it when I'm back from vacation in the second week of Aug.

**Comment by gasche** [comment]
Thanks for the update, and have nice vacations :-)

**Comment by gasche** [comment]
(gentle ping)

**Comment by gasche** [comment]
Would maybe someone else be interested in looking at this? @nojb, @Octachron, @trefis?

File: .gitattributes [hunk 3]
```diff
@@ -16,8 +16,8 @@
 # Default behaviour, for if core.autocrlf isn't set
 * text=auto
 
-# Don't believe there's a way to wrap lines in .gitattributes
-.gitattributes ocaml-typo=long-line
+# Don't believe there's a way to wrap lines in .gitattributes files
```
Comment (dra27) [inline .gitattributes]: You could make this stronger - there is no way, nor can there ever be, given backwards compatibility

Comment (dra27) [reply .gitattributes]: (it would be an amusing diversion, for some value of amusing, to lint these files such that *comment blocks* must always be 80 chars or fewer, but that "code" blocks may exceed it)

File: .gitattributes [hunk 4]
```diff
@@ -38,68 +38,81 @@
...
-
-/.mailmap                ocaml-typo=long-line,missing-header,non-ascii
-/.merlin                 ocaml-typo=missing-header
-/Changes                 ocaml-typo=utf8,missing-header
-/INSTALL                 ocaml-typo=missing-header
-/LICENSE                 ocaml-typo=long-line,very-long-line,missing-header
+*.md                     typo.missing-header
+README*                  typo.missing-header
+*.adoc                   typo.missing-header
+*.adoc                   typo.long-line=may
```
Comment (dra27) [inline .gitattributes:44]: Can the two `*.adoc` lines be merged? It should technically be `*.adoc typo.long-line=may typo.missing-header=may` but I agree with the change where the handling of `missing-header` is concerned.

File: .gitattributes [hunk 5]
```diff
@@ -38,68 +38,81 @@
...
+.depend                  typo.prune
 
-asmcomp/*/emit.mlp       ocaml-typo=tab,long-line,unused-prop
-asmcomp/power/NOTES.md   ocaml-typo=missing-header
+# Makefiles may contain tabs
+Makefile*                typo.tab=may
 
-/config/gnu              ocaml-typo=prune
+asmcomp/*/emit.mlp       typo.tab=may typo.long-line=may
+asmcomp/power/NOTES.md   typo.missing-header
```
Comment (dra27) [inline .gitattributes]: I can't see a problem with `*.md typo.missing-header` with the `*.adoc` pattern above?

File: .gitattributes [hunk 6]
```diff
@@ -38,68 +38,81 @@
...
-testsuite/tools/*.asm                                   ocaml-typo=missing-header
-testsuite/typing                                        ocaml-typo=missing-header
+stdlib/hashbang     typo.white-at-eol typo.missing-lf
+
+testsuite/tests/**                                      typo.missing-header
+testsuite/tests/lib-bigarray-2/bigarrf.f                typo.tab
+testsuite/tests/lib-unix/win-stat/fakeclock.c           typo.missing-header=false
+testsuite/tests/misc-unsafe/almabench.ml                typo.long-line
+testsuite/tests/tool-toplevel/strings.ml                typo.utf8
+testsuite/tests/typing-unboxed-types/test.ml.reference* typo.white-at-eof
```
Comment (dra27) [inline .gitattributes]: This can be removed, given the new prune on reference files below

File: .gitattributes [hunk 7]
```diff
@@ -38,68 +38,81 @@
...
+testsuite/tests/typing-modules/firstclass.ml typo.long-line
+testsuite/tests/typing-gadts/test.ml         typo.long-line
+testsuite/tests/typing-gadts/pr7160.ml       typo.long-line
+testsuite/tests/typing-modules/pr7726.ml     typo.long-line
 
-tools/magic         ocaml-typo=missing-header
+tools/magic         typo.missing-header
 
-yacc/*.[ch]   ocaml-typo=long-line,very-long-line,unused-prop
+yacc/*.[ch]         typo.long-line typo.very-long-line=may
```
Comment (dra27) [inline .gitattributes]: Shouldn't this be `typo.long-line=may` too?

File: .gitattributes [hunk 8]
```diff
@@ -38,68 +38,81 @@
...
+emacs/COPYING            typo.tab typo.non-printing typo.missing-header
+emacs/ocamltags.in       typo.non-printing
 
-/manual                  ocaml-typo=prune
+/experimental            typo.prune
 
-ocamldoc/Changes.txt     ocaml-typo=missing-header
-ocamldoc/ocamldoc.sty    ocaml-typo=missing-header
-ocamldoc/odoc_index.html ocaml-typo=missing-header
+/manual                  typo.prune
```
Comment (dra27) [inline .gitattributes:74]: Can this have a comment above it explaining that this is a temporary state of affairs (i.e. make it clear to a willing volunteer that it would be preferred to bring the manual into the arms of `check-typo` as well)?

File: .gitattributes [hunk 9]
```diff
@@ -38,68 +38,81 @@
...
 
-otherlibs/win32unix/readlink.c    ocaml-typo=long-line
-otherlibs/win32unix/stat.c        ocaml-typo=long-line
-otherlibs/win32unix/symlink.c     ocaml-typo=long-line
+ocamldoc/**              typo.long-line=may
```
Comment (dra27) [inline .gitattributes:76]: Similarly - is this another attribute that it would be worth admitting our shame of?

File: .gitattributes [hunk 10]
```diff
@@ -38,68 +38,81 @@
...
+testsuite/tests/typing-gadts/pr7160.ml       typo.long-line
+testsuite/tests/typing-modules/pr7726.ml     typo.long-line
 
-tools/magic         ocaml-typo=missing-header
+tools/magic         typo.missing-header
 
-yacc/*.[ch]   ocaml-typo=long-line,very-long-line,unused-prop
+yacc/*.[ch]         typo.long-line typo.very-long-line=may
 
 # Line-ending specifications, for Windows interoperability
```
Comment (dra27) [inline .gitattributes:123]: I like the separation of lint attributes and line-ending controls. Given the overall changes to this file, can we restructure so that all lint instructions (except, perhaps, the one for `.gitattributes` as a very special case) are at the bottom. So can this point onwards be moved below the `Changes` merge driver comment on line 38?

File: .gitattributes [hunk 11]
```diff
@@ -38,68 +38,81 @@
...
 # tools/ci/appveyor/appveyor_build.cmd only has missing-header because
 # dra27 too lazy to update check-typo to interpret Cmd-style comments!
-/tools/ci/appveyor/appveyor_build.cmd      ocaml-typo=long-line,very-long-line,missing-header,non-ascii text eol=crlf
-/tools/ci/appveyor/appveyor_build.sh       ocaml-typo=non-ascii
+/tools/ci/appveyor/appveyor_build.cmd       typo.long-line typo.very-long-line typo.missing-header typo.non-ascii
+/tools/ci/appveyor/appveyor_build.sh        typo.non-ascii
+/tools/ci/inria/remove-sinh-primitive.patch typo.white-at-eol typo.missing-header
 
+# ignore auto-generated .depend files
+.depend                  typo.prune
```
Comment (dra27) [inline .gitattributes:59]: This needs to be `.depend*`, or you need to an entry for `.depend.menhir` (the original pattern in check-typo was `.depend*` which would have been to catch `.depend.nt` files as well, I expect)

File: tools/check-typo [hunk 12]
```diff
@@ -35,27 +35,9 @@
...
 # The rule names are the ones shown above in parentheses.
 
-# Built-in exceptions:
+# Built-in exception:
 # - Any file git identifies as binary
 #   is automatically exempt from all the rules.
-# - Any file whose name matches one of the following patterns is
-#   automatically exempt from all rules
-#     *.reference
-#     *.opt_reference
```
Comment (dra27) [inline tools/check-typo:44]: Just FTR, `*.opt_reference` was removed with the ocamltest migration (gone since 8c2d90543fce1e5aca1bd9bc796a51f04be48b4f)

File: .gitattributes [hunk 13]
```diff
@@ -38,68 +38,81 @@
...
+testsuite/tests/tool-toplevel/strings.ml                typo.utf8
+testsuite/tests/typing-unboxed-types/test.ml.reference* typo.white-at-eof
+testsuite/tests/win-unicode/*.ml                        typo.utf8
+testsuite/tools/*.S                                     typo.missing-header
+testsuite/tools/*.asm                                   typo.missing-header
+testsuite/typing                                        typo.missing-header
+
+# prune testsuite reference files
+testsuite/tests/**/*.reference               typo.prune
+testsuite/tests/**/reference                 typo.prune
```
Comment (dra27) [inline .gitattributes]: This is used by *just* the `lib-scanf-2` test - can't we just rename the file and get rid of the special case?

File: tools/check-typo [hunk 14]
```diff
@@ -35,27 +35,9 @@
...
-#     */.depend*
-# - Any file whose name begins with "Makefile" is automatically exempt
-#   from the "tabs" rule.
-# - Any file whose name matches one of the following patterns is
-#   automatically exempt from the "missing-header" rule.
-#     *.mlpack
-#     *.mllib
-#     *.mltop
-#     *.odocl
-#     *.clib
```
Comment (dra27) [inline tools/check-typo:55]: Again, just FTR: these were ocamlbuild special cases

File: tools/check-typo [hunk 15]
```diff
@@ -88,13 +70,42 @@
...
+# for example,
+#   typo.long-line: set
+#   typo.missing-header: may
+#   typo.very-long-line: false
+# should result in "long-line missing-header?"
+get_attrs() {
+  env $OCAML_CT_GIT_INDEX git check-attr --all $OCAML_CT_CA_FLAG "$1" \
+  | grep -o " typo\\..*$" | sed "s/ typo\\.//g" \
+  | grep -v ": unset" | grep -v ": false" \
+  | sed "s/: set//g" | sed "s/: true//g" | sed "s/: may/?/g"
```
Comment (dra27) [inline tools/check-typo:98]: This pipe can be more simply (or at least compactly): `sed -ne '/: \(set\|true\|may\)/s/.* typo\.//p' -e 's/: \(set\|true\)//' -e 's/:.*/?/'`

File: tools/check-typo [hunk 16]
```diff
@@ -88,13 +70,42 @@
...
+}
+
 # Special case for recursive call from the find command (see IGNORE_DIRS).
 case "$1" in
   --check-prune)
-    case `env $OCAML_CT_GIT_INDEX git check-attr $OCAML_CT_CA_FLAG \
-                                                 ocaml-typo "$2" 2>/dev/null` in
-      *prune*) echo "INFO: pruned directory $2 (ocaml-typo=prune)" >&2; exit 0;;
-      *) exit 3;;
+    case $2,"$(get_attrs "$2" | grep prune)" in
```
Comment (dra27) [inline tools/check-typo]: Pedantically, do `grep -xF prune`

File: tools/check-typo [hunk 17]
```diff
@@ -88,13 +70,42 @@
...
+      *,*prune*|.git,*)
```
Comment (dra27) [inline tools/check-typo]: The exact line should be `prune` - this is brittle if we happened to add another attribute containing `prune` as a substring? If the grep flags I suggest (text and wholeline) are added, I think the `*`s can go from `*prune*`

**Review by dra27 — COMMENTED** [review]
Sorry for the delay reviewing this. I have reviewed very carefully the changes to `.gitattributes` and reckon that all the pertinent logic which was hard-coded in `check-typo` is now transferred to `.gitattributes`.

The changes to the script itself look good to me - perhaps @damiendoligez could cast a quick eye over it for any concerns about coding style?

**Review by dra27 — COMMENTED** [review]
Sorry for the delay reviewing this. I have reviewed very carefully the changes to `.gitattributes` and reckon that all the pertinent logic which was hard-coded in `check-typo` is now transferred to `.gitattributes`.

The changes to the script itself look good to me - perhaps @damiendoligez could cast a quick eye over it for any concerns about coding style?

File: .gitattributes (hunk 1 above)
Comment (damiendoligez) [reply .gitattributes:20]: @dra27 in my eye, the typo checks are much weaker than what a proper lint tool would do.

File: .gitattributes (hunk 2 above)
Comment (damiendoligez) [reply .gitattributes]: @dra27 I thought it was doing that already? Did I just dream it?

File: .gitattributes [hunk 18]
```diff
@@ -38,68 +38,81 @@
...
-emacs/*.el               ocaml-typo=long-line,unused-prop
-emacs/caml.el            ocaml-typo=long-line,unused-prop,missing-header
-emacs/COPYING            ocaml-typo=tab,non-printing,missing-header
-emacs/ocamltags.in       ocaml-typo=non-printing
+/config/gnu              typo.prune
 
-/experimental            ocaml-typo=prune
+emacs/*.el               typo.long-line=may
+emacs/caml.el            typo.long-line=may typo.missing-header
+emacs/COPYING            typo.tab typo.non-printing typo.missing-header
```
Comment (damiendoligez) [inline .gitattributes]: This one should be `typo.prune`: we're not ever going to change the GPL text, even if we invent new typo checks or whatever.

File: .gitattributes (hunk 13 above)
Comment (damiendoligez) [reply .gitattributes]: Seconded. That should slightly simplify the test script itself.

**Review by damiendoligez — APPROVED** [review]
Looks good!

File: .gitattributes (hunk 2 above)
Comment (dra27) [reply .gitattributes]: No - you definitely need both; it could be done by having the test for `long-line` being between 81 and ~131~132 characters and then `very-long-line` being > 132.

Comment (damiendoligez) [reply .gitattributes]: That sounds like a good idea (modulo the off-by-one error in your description).

Comment (dra27) [reply .gitattributes]: Oops!

Comment (gasche) [reply .gitattributes]: This is a bit tricky (as in: it requires me to rediscover `awk` again). Given the large latency on my side, I won't touch this feature request and leave it for a future PR.

File: .gitattributes (hunk 1 above)
Comment (gasche) [reply .gitattributes:20]: (Out of laziness, let's keep the current name for now.)

File: .gitattributes (hunk 4 above)
Comment (gasche) [reply .gitattributes:44]: I prefer to keep them separate as the first part is part of a "no header for text files" block, while the second is more .adoc-specific.

File: .gitattributes (hunk 3 above)
Comment (gasche) [reply .gitattributes]: I changed it. Note that I think we could now fit within a 80-characters budget, given that the new typo-handling style supports breaking a line by duplicating the path part. But we are still limited by the path length, which reaches 67 for some testsuite files and could conceivably go above 80, so I kept the comment.

File: .gitattributes (hunk 5 above)
Comment (gasche) [reply .gitattributes]: Indeed, thanks!

File: .gitattributes (hunk 13 above)
Comment (gasche) [reply .gitattributes]: Done.

File: tools/check-typo (hunk 15 above)
Comment (gasche) [reply tools/check-typo:98]: My sed-fu is not high enough to understand your version, so I'll stick with dumb pipes for now.

File: tools/check-typo (hunk 16 above)
Comment (gasche) [reply tools/check-typo]: I just checked that [FreeBSD](https://www.freebsd.org/cgi/man.cgi?query=grep&sektion=&n=1) and [OSX](https://ss64.com/osx/grep.html) both support `-x` and its longer form `--line-regexp`, so I will use that, thanks.

**Comment by dra27** [comment]
I just ran the script over the entire repo and found that the `yacc/*.[ch]` was triggering unused warnings (because it was `typo.long-line` rather than `typo.long-line=may`).

I've also pushed a commit which makes it so that `typo.very-long-line` implies `typo.long-line=may`.

This needs rebasing, so feel free to squash the first commit and potentially drop the latter commit when you do!

File: tools/check-typo (hunk 16 above)
Comment (gasche) [reply tools/check-typo]: (I'm currently confused as to whether `get_attrs` really returns only the value, or may carry some may/must information also on the line.)

**Comment by gasche** [comment]
I rebased the PR and I'm in the process of going over the review comments. I hope to finish today to get this PR dealt with.

**Comment by gasche** [comment]
@dra27 I'm sorry, I am just seeing that you did pushes on your own. (I force-pushed when I rebased.) I fixed the yacc-stuff as well, but I'm happy to integrate the long-line/very-long-line commit in the PR, could you push it again?

**Comment by dra27** [comment]
No apology needed - it's your branch :slightly_smiling_face: Commit re-pushed...

File: tools/check-typo (hunk 16 above)
Comment (gasche) [reply tools/check-typo]: (The may/must information is encoded as a question mark after the attribute name, so indeed `-x` is correct.)

**Comment by gasche** [comment]
I believe that I should now have addressed all review comments (thanks!).

File: .gitattributes (hunk 6 above)
Comment (dra27) [reply .gitattributes]:
```suggestion
```

**Review by dra27 — COMMENTED** [review]
One tweak - though mainly a chance to use GitHub's new suggest edit feature :)

**Comment by gasche** [comment]
@dra27: the way I read the diff produced by this indeed-intriguing feature, you are propose to remove the `reference.ml* white-at-eof` line. But if I remove it on my local copy, then I do get an alarm:

```
$ ./tools/check-typo testsuite/tests/typing-unboxed-types/test.ml.reference*
testsuite/tests/typing-unboxed-types/test.ml.reference-flat:213.1: [white-at-eof] empty line(s) at EOF
testsuite/tests/typing-unboxed-types/test.ml.reference-noflat:179.1: [white-at-eof] empty line(s) at EOF
```

What's going on here?

File: .gitattributes (hunk 9 above)
Comment (damiendoligez) [reply .gitattributes:76]: There are so many long lines in ocamldoc that I doubt anyone will ever fix it (except by releasing a tool that obsoletes ocamldoc, wink, wink).

File: .gitattributes (hunk 6 above)
Comment (damiendoligez) [reply .gitattributes]: @dra27 did you miss the `*` at the end of the pattern?

File: testsuite/tests/lib-scanf-2/tscanf2_master.ml [hunk 19]
```diff
@@ -2,8 +2,8 @@
 
 include unix
 modules = "tscanf2_io.ml"
-files = "tscanf2_slave.ml"
-reference = "${test_source_directory}/reference"
+files = "tscanf2_worker.ml"
```
Comment (damiendoligez) [inline testsuite/tests/lib-scanf-2/tscanf2_master.ml:5]: These changes are completely unrelated to check-typo (unless you intend to add a style checker). They should be in a separate PR (and probably touch many more files).

**Comment by gasche** [comment]
In the interest of not waiting too much on this issue (its discussion in https://github.com/ocaml/ocaml/pull/2096#issuecomment-430376025 bumped it up in my priority list), I'm thinking of merging this PR today or tomorrow.

**Review by damiendoligez — APPROVED** [review]
Looks OK except for the unrelated changes.

File: .gitattributes (hunk 6 above)
Comment (dra27) [reply .gitattributes]: You're right, I had! In which case, it should be `typo.prune`, but perhaps this can be simplified by altering the tests. The following files:

```
tests/typing-unboxed-types/test.ml.reference-flat
tests/typing-unboxed-types/test.ml.reference-noflat
tests/translprim/array_spec.compilers.reference.flat
tests/translprim/module_coercion.compilers.reference.flat
tests/translprim/module_coercion.compilers.reference.no-flat
tests/translprim/array_spec.compilers.reference.no-flat
```
are the only ones which don't end `.reference`. Perhaps they should just be renamed to `.flat.reference` and `.no-flat.reference` so that the normal prune rule comes into play?

**Comment by shindere** [comment]
David Allsopp (2018/10/18 07:03 -0700):
> You're right, I had! In which case, it should be `typo.prune`, but perhaps this can be simplified by altering the tests. The following files:
> 
> ```
> tests/typing-unboxed-types/test.ml.reference-flat
> tests/typing-unboxed-types/test.ml.reference-noflat
> tests/translprim/array_spec.compilers.reference.flat
> tests/translprim/module_coercion.compilers.reference.flat
> tests/translprim/module_coercion.compilers.reference.no-flat
> tests/translprim/array_spec.compilers.reference.no-flat
> ```
> are the only ones which don't end `.reference`. Perhaps they should
> just be renamed to `.flat.reference` and `.no-flat.reference` so that
> the normal prune rule comes into play?

Yes, please!

Sorry for having done things the wrong way initially!

**Comment by gasche** [comment]
I'm busy with other things and would like to get done with this PR. Is it possible to leave those further suggestions to future PRs?

**Comment by shindere** [comment]
Gabriel Scherer (2018/10/18 07:25 -0700):
> I'm busy with other things and would like to get done with this PR. Is
> it possible to leave those further suggestions to future PRs?

Fine with me :)

**Comment by damiendoligez** [comment]
@gasche just let me push a commit to rename the reference files before you merge.

**Comment by damiendoligez** [comment]
Done.

**Comment by gasche** [comment]
Thanks! Merged.
